File: Residential_Controller.py
# -------------- ELEVATORS ----------------------------
class Elevator:
  def __init__(self, id):
    self.id = id
    self.direction = 'idle'
    self.CurrentFloor = 1
    self.Doors = "close"
    self.FloorRequestButtonList = []

  def checkList(self):
    while self.FloorRequestButtonList != 0:


      self.FloorRequestButtonList.pop([0])

  def move(self,requestedFloor):

    if self.CurrentFloor < requestedFloor:
      while self.CurrentFloor < requestedFloor:
        # print(self.CurrentFloor)
        self.CurrentFloor += 1

      # print('Elevator ' + String(self.id) + ' is at floor ' + String(self.CurrentFloor))

    if self.CurrentFloor > requestedFloor:
      while self.CurrentFloor > requestedFloor:
        self.CurrentFloor = self.CurrentFloor - 1

      # print('Elevator ', str(id) ,
      #       ' is at floor ', str(self.CurrentFloor))

  def addFloorRequestButtonList(self, floor):
    self.FloorRequestButtonList.append(floor)

# -------------- OUTSIDE BUTTON ----------------
class CallButton:
  def __init__(self, direction, floor):
    self.direction = direction
    self.floor = floor
    self.Light = "off"

# --------------INSIDE BUTTON----------------
class InsideButton:
  def __init__(self, Floor):
    self.Floor = Floor
    self.Light = "off"
# --------------COLUMN-----------------------
class Column:
  def __init__(self, ID, NbElevator, NbFloor):
    self.ID = ID
    self.NbElevator = NbElevator
    self.NbFloor = NbFloor
    self.elevatorsList = []
    self.CallButtonList = []
    self.buttonList = []

    # nombre elevator
    for i in range(NbElevator):
      elevator = Elevator(NbFloor)
      self.elevatorsList.append(elevator)

    #  bouton Up and Down for i in range(NbFloor):
    for i in range(NbFloor):
      if i != 1:
        callButton = CallButton("down", i)
        self.CallButtonList.append(callButton)

      if i != NbFloor:
        callButton = CallButton("up", i)
        self.CallButtonList.append(callButton)


  def requestElevator(self,direction, requestedFloor):
    bestElevator = FindElevator(direction, requestedFloor)

    bestElevator.move(requestedFloor)
    return bestElevator

  def requestFloor(self,requestedFloor, elevator):
    elevator.addFloorRequestButtonList(requestedFloor)
    elevator.checkList()
  

  def FindElevator(self,direction, requestedFloor):
    bestElevator = None
    bestCase = None
    print('find elevator')

    for elevator in self.elevatorsList:
      print('for each')

      if requestedFloor == elevator.position and direction == elevator.direction:
        print('bestCase ', bestCase)
        if bestCase == None or bestCase > 1:
          print("1")
          print(elevator)
          bestCase = 1
          bestElevator = elevator
        
        
      elif requestedFloor == elevator.position and elevator.direction == 'idle':
        print('bestCase ', bestCase)
        if (bestCase == None or bestCase > 2):
          print("2")
          print(elevator)
          bestCase = 2
          bestElevator = elevator
        
      elif (requestedFloor > elevator.position and elevator.direction == "idle"):

        print('bestCase ', bestCase)
        if bestCase == None or bestCase > 3:
          print("3")
          print(elevator)
          bestCase = 3
          bestElevator = elevator
          # self.NearestElevator(requestedFloor)
        
      elif (requestedFloor < elevator.position or elevator.direction == "idle"):
        print('bestCase ', bestCase)

        if (bestCase == None or bestCase > 4):
          print("4")
          print(elevator)
          bestCase = 4
          bestElevator = elevator
          # self.NearestElevator(requestedFloor)

      
      elif (requestedFloor > elevator.position or elevator.direction == direction):
        print('bestCase ', bestCase)

        if (bestCase == None or bestCase > 5):
          print("5")
          print(elevator)
          bestCase = 5
          bestElevator = elevator
          # self.NearestElevator(requestedFloor)
        
      elif (requestedFloor < elevator.position and elevator.direction == direction):
        print('bestCase ', bestCase)

        if (bestCase == None or bestCase > 6):
          print("6")
          print(elevator)
          bestCase = 6
          bestElevator = elevator
          # self.NearestElevator(requestedFloor)
      elif (elevator.direction == "idle"):
        print('bestCase ', bestCase)

        if (bestCase == None or bestCase > 7):
          print("7")
          print(elevator)
          bestCase = 7
          bestElevator = elevator
          # self.NearestElevator(requestedFloor)

    print('bestElevator ', bestElevator)
    return bestElevator

# column1 = new Column(1, 2, 10)
# # print(column1)

# CallButton1 = new CallButton("up", 6)
# print(CallButton1)

# column1.FindElevator(1, 5, 5)
# print(column1.FindElevator)

# elevator1 = new Elevator(1,5,10)
# print('FloorRequestButtonList', elevator1.FloorRequestButtonList)

# InsideButton1 = new InsideButton(8)
# print(InsideButton1)


# class column parameters --> (ID, NbElevator, NbFloor)
column1 = Column(1, 2, 10)

# elevators(id, nbFloor, position, direction, FloorRequestButtonList)
column1.elevatorsList[0].CurrentFloor == 1
column1.elevatorsList[1].CurrentFloor == 2

# user outside
quelAscenseurTuMenvois = column1.requestElevator('up', 5)

# user inside
# column1.requestFloor(10, quelAscenseurTuMenvois)

File: Residential_Controller.js
//----------------------- Column controling elevators ------------------

class Column {

  constructor(ID, NbElevator, NbFloor) {
    this.ID = ID;
    this.NbElevator = NbElevator
    this.NbFloor = NbFloor
    this.elevatorsList = []
    this.CallButtonList = []
    this.buttonList = []


    // nombre elevator

    for (var i = 0; i < NbElevator; i++) {
      var elevator = new Elevator(i, NbFloor)
      elevator.id = i
      elevator.CurrentFloor = 0
      this.elevatorsList.push(elevator);
    }

    //  bouton Up and Down

    for (var i = 1; i < NbFloor + 1; i++) {
      if (i != 1) {
        var callButton = new CallButton("down", i);
        this.CallButtonList.push(callButton)
      }
      if (i != NbFloor) {
        var callButton = new CallButton("up", i)
        this.CallButtonList.push(callButton)
      }
    }
  }

  // user call elevator (methods chaining)
  requestElevator(direction, requestedFloor) {
    var bestElevator = this.FindElevator(direction, requestedFloor)

    bestElevator.move(requestedFloor)
    return bestElevator
  }

  requestFloor(requestedFloor, elevator) {
    elevator.addFloorRequestButtonList(requestedFloor)
    elevator.checkList()
  }

  FindElevator(direction, requestedFloor) {


    var bestNearestElevatorGap = 10000000;
    var bestElevator;
    var bestCase = null;

    this.elevatorsList.forEach((elevator) => {

        if (requestedFloor == elevator.CurrentFloor && direction == elevator.direction) {
          if (bestCase == null || bestCase > 1) {
            console.log("1")
            console.log(elevator)
            bestCase = 1
            bestElevator = elevator
          }
        } else if (requestedFloor == elevator.CurrentFloor && elevator.direction == 'idle') {
          if (bestCase == null || bestCase > 2) {
            console.log("2")
            console.log(elevator)
            bestCase = 2
            bestElevator = elevator
          }
        } else if (requestedFloor > elevator.CurrentFloor && elevator.direction == "idle") {

          if (bestCase == null || bestCase > 3) {
            console.log("jsuisla")
            bestCase = 3
            console.log(bestCase)
            console.log(elevator)
            bestElevator = elevator

            // this.NearestElevator(requestedFloor)
          } else if (bestCase == 3) {
            var gap = Math.abs(elevator.CurrentFloor - requestedFloor)
            if (bestNearestElevatorGap >= gap) {
              bestElevator = elevator
              bestNearestElevatorGap = gap


            }
          }
        } else if (requestedFloor < elevator.CurrentFloor && elevator.direction == "idle") {

          if (bestCase == null || bestCase > 4) {
            console.log("4")
            console.log(elevator)
            bestCase = 4
            bestElevator = elevator
            // this.NearestElevator(requestedFloor)

          } else if (bestCase == 4) {
            var gap = Math.abs(elevator.CurrentFloor - requestedFloor);
            if (bestNearestElevatorGap >= gap) {
              bestElevator = elevator
              bestNearestElevatorGap = gap
            }
          }
        } else if (requestedFloor > elevator.CurrentFloor && elevator.direction == direction) {

          if (bestCase == null || bestCase > 5) {
            console.log("5")
            console.log(elevator)
            bestCase = 5
            bestElevator = elevator
            // this.NearestElevator(requestedFloor)
          } else if (bestCase == 5) {
            var gap = Math.abs(elevator.CurrentFloor - requestedFloor);
            if (bestNearestElevatorGap >= gap) {
              bestElevator = elevator
              bestNearestElevatorGap = gap
            }
          }
        } else if (requestedFloor < elevator.CurrentFloor && elevator.direction == direction) {

          if (bestCase == null || bestCase > 6) {
            console.log("6")
            console.log(elevator)
            bestCase = 6
            bestElevator = elevator
            // this.NearestElevator(requestedFloor)
          } else if (bestCase == 6) {
            var gap = Math.abs(elevator.CurrentFloor - requestedFloor);
            if (bestNearestElevatorGap >= gap) {
              bestElevator = elevator
              bestNearestElevatorGap = gap
            }
          }
        } else if (elevator.direction == "idle") {

          if (bestCase == null || bestCase > 7) {
            console.log("7")
            console.log(elevator)
            bestCase = 7
            bestElevator = elevator
            // this.NearestElevator(requestedFloor)
          } else if (bestCase == 7) {
            var gap = Math.abs(elevator.CurrentFloor - requestedFloor);
            if (bestNearestElevatorGap >= gap) {
              bestElevator = elevator
              bestNearestElevatorGap = gap
            }
          }


        }
      })
      console.log('la best elevator', bestElevator)
      return bestElevator
  }

}


//-------------- user calling elevator from his floor ----------------
class CallButton {
  constructor(direction, floor) {
    this.direction = direction;
    this.floor = floor;
    this.Light = "off";
  }
}

//-------------- user inside elevator----------------
class InsideButton {
  constructor(Floor) {
    this.Floor = Floor;
    this.Light = "off";
  }

}

//-------------- Elevator ----------------------------
class Elevator {
  constructor(id) {
    this.id = id
    this.direction = 'idle'
    this.CurrentFloor = 0
    this.Doors = "close"
    this.FloorRequestButtonList = []

  }

  checkList() {
    while (this.FloorRequestButtonList != 0) {

      this.move(this.FloorRequestButtonList[0])

      this.FloorRequestButtonList.pop([0])


    }

  }

  move(requestedFloor) {

    if (this.CurrentFloor < requestedFloor) {
      while (this.CurrentFloor < requestedFloor) {
        // console.log(this.CurrentFloor)
        this.CurrentFloor += 1;

        console.log('Elevator ' + String(this.id) + ' is at floor ' + String(this.CurrentFloor))
      }
    }

    if (this.CurrentFloor > requestedFloor) {
      while (this.CurrentFloor > requestedFloor) {
        this.CurrentFloor = this.CurrentFloor - 1;
        console.log('Elevator ' + String(this.id) + ' is at floor ' + String(this.CurrentFloor))

      }
    }
  }

  addFloorRequestButtonList(floor) {
    this.FloorRequestButtonList.push(floor)
  }


  // user request floor inside

}

// class column parameters --> (ID, NbElevator, NbFloor)
column1 = new Column(1, 2, 10)

// elevators(id, nbFloor, position, direction, FloorRequestButtonList)
column1.elevatorsList[0].CurrentFloor = 2
column1.elevatorsList[1].CurrentFloor = 7

console.log('Elevator!!!')
console.log(column1.elevatorsList[0])

// user outside
quelAscenseurTuMenvois = column1.requestElevator('down', 6)
console.log(quelAscenseurTuMenvois)